torch/fx/experimental/proxy_tensor.py

```diff
@@ -70,7 +72,64 @@ def decompose(decomposition_table):
 proxy_slot = object()
 no_default = object()
 
+
 py_sym_types = (SymInt, SymFloat, SymBool)
+
+
+class SuppressGuardsAndCheckIfAnyAdded:
+    def __init__(self, shape_env):
+        self.guards_added = False
+        self.shape_env = shape_env
+        self.num_guards = len(shape_env.guards)
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, *args):
+        self.guards_added = len(self.shape_env.guards) != self.num_guards
+        self.shape_env.guards = self.shape_env.guards[0:self.num_guards]
+
+@dataclass
+class SymExprHash:
+    sym_obj: py_sym_types
+
+    def __hash__(self) -> int:
+        return hash((type(self.sym_obj), hash(self.sym_obj.node.expr)))
+
+    def __eq__(self, value: "SymExprHash") -> bool:
+        assert isinstance(value, SymExprHash)
+
+        sg = SuppressGuardsAndCheckIfAnyAdded(self.sym_obj.node.shape_env)
+        with sg:
+            eq = definitely_true(self.sym_obj == value.sym_obj)
+
+        return eq and not sg.guards_added
```

Comment:
This is a weird way to do it - I thought we agreed to check the underlying expr, and assume that identical expr means its sound to follow a caching pattern and produce an identical symnode, instead of doing all this guard business?

Reply:
Can you explain why the guard business is not sound ? im happy to switch but you havent given any details of why this incorrect.

Reply:
1) Design complaint: guards are a notion internal to executing shape_env operations - you have no idea how many other downstream systems are using the view of the guards as they are added. It's very onerous to go from an append-only system to a system that can remove elements. 

2) Correctness complaint: There's a lot of stuff that can happen when evaluating an op that *does not* incur a guard but accumulates some stored entities in a structure that changes which guards are installed. I dont know all the things off the top of my head - but let's look at a simple example. Imagine you run an operation that introduces a new symbol under SuppressGuardsAndCheckIfAnyAdded. A guard will not be added there, but we will have a value in val_to_var forever, as you are not unrolling shape_env state, just trimming guards. All downstream systems will behave as if that symbol is there, when in reality, it was only used for some speculation. The API of SuppressGuardsAndCheckIfAnyAdded does not actually protect you from side effects, it is insufficient to prove that running ops under it does not change any underlying structures.